core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableSortedAggregate.java

```diff
@@ -90,6 +101,133 @@ public EnumerableSortedAggregate(
+      aggs.add(new AggImpState(call.i, call.e, false));
+    }
+
+    // Function0<Object[]> accumulatorInitializer =
+    //     new Function0<Object[]>() {
+    //         public Object[] apply() {
+    //             return new Object[] {0, 0};
+    //         }
+    //     };
+    final List<Expression> initExpressions = new ArrayList<>();
+    final BlockBuilder initBlock = new BlockBuilder();
+
+    final List<Type> aggStateTypes = createAggStateTypes(
+        initExpressions, initBlock, aggs, typeFactory);
+
+    final PhysType accPhysType =
+        PhysTypeImpl.of(typeFactory,
+            typeFactory.createSyntheticType(aggStateTypes));
+
+    declareParentAccumulator(initExpressions, initBlock, accPhysType);
+
+    final Expression accumulatorInitializer =
+        builder.append("accumulatorInitializer",
+            Expressions.lambda(
+                Function0.class,
+                initBlock.toBlock()));
+
+    // Function2<Object[], Employee, Object[]> accumulatorAdder =
+    //     new Function2<Object[], Employee, Object[]>() {
+    //         public Object[] apply(Object[] acc, Employee in) {
+    //              acc[0] = ((Integer) acc[0]) + 1;
+    //              acc[1] = ((Integer) acc[1]) + in.salary;
+    //             return acc;
+    //         }
+    //     };
+    final ParameterExpression inParameter =
+        Expressions.parameter(inputPhysType.getJavaRowType(), "in");
+    final ParameterExpression acc_ =
+        Expressions.parameter(accPhysType.getJavaRowType(), "acc");
+
+    createAccumulatorAdders(
+        inParameter, aggs, accPhysType, acc_, inputPhysType, builder, implementor, typeFactory);
+
+    final ParameterExpression lambdaFactory =
+        Expressions.parameter(AggregateLambdaFactory.class,
+            builder.newName("lambdaFactory"));
+
+    implementLambdaFactory(builder, inputPhysType, aggs, accumulatorInitializer,
+        false, lambdaFactory);
+
+    final BlockBuilder resultBlock = new BlockBuilder();
+    final List<Expression> results = Expressions.list();
+    final ParameterExpression key_;
+    final Type keyType = keyPhysType.getJavaRowType();
+    key_ = Expressions.parameter(keyType, "key");
+    for (int j = 0; j < groupCount; j++) {
+      final Expression ref = keyPhysType.fieldReference(key_, j);
+      results.add(ref);
+    }
+
+    for (final AggImpState agg : aggs) {
+      results.add(
+          agg.implementor.implementResult(agg.context,
+              new AggResultContextImpl(resultBlock, agg.call, agg.state, key_,
+                  keyPhysType)));
+    }
+    resultBlock.add(physType.record(results));
+
+    final Expression keySelector_ =
+        builder.append("keySelector",
+            inputPhysType.generateSelector(parameter,
+                groupSet.asList(),
+                keyPhysType.getFormat()));
+    // Generate the appropriate key Comparator.
+    final Expression comparator = keyPhysType.generateComparator(getTraitSet().getCollation());
```

Comment:
I will need to double check and probably revise this part to make KeySelector match with KeyComparator.

Reply:
In fact. I cannot find a way to turn on top down opt by Hook.PLANNER